Initial interpreter

pmer opened the pull request:
This PR contains an initial parser and interpreter for Stratagem. It does not contain a type checker.

## Changes to the formal language

- Added function return type annotation.
- Added a unit type. This type has only one value `()` and its type is written `()`. For the most part it takes the place of `null` from FWJS. It's the type returned by an empty program. We should be able to remove this type if we want to say that's an invalid construct and that a program must contain at least one expression.
- Added a sequence expression: `e; e`. Its value is the value of the second `e`.

## Changes to the code base

- [x] Addition of the `let` expression and of the binary operator `!=`.
- [x] Addition of the String literal.
- [x] Removed the concept of a statement. Stratagem is an expression-oriented language.
- [x] Added `seq` expression to grammar. A `seq` is a list of one or more `expr` and cannot end with a separator (`;`). The `prog` is now just a single `seq`.
- [x] SeqExpr holds a `List<Expression>` instead of two `Expression`s called `e1` and `e2`.
- [x] UnitVal replaces NullVal.
- [x] Remove of the `if` with no `else` clause, the `while`, and the `print` expressions. Although... removing `print` might be silly... :)
- [x] Added syntax for types but they have no semantics yet.
- [x] New `stratagemScripts` directory that we can populate with interesting Stratagem programs. There is one program so far.
- [ ] Addition of ternary operator.
- [ ] Functions can take multiple arguments.
- [x] Updated dependencies: ANTLR goes to 4.6

## Type checker

No type checker in this PR. That will come soon.

pmer committed:
Add README

pmer committed:
Fix typo in comment

pmer committed:
Create gitignore

pmer committed:
libs: Update ANTLR to 4.6

pmer committed:
grammar: Initial Stratagem language

pmer committed:
ExprBuilder: Remove excess whitespace

pmer committed:
ExprBuilder: A program is a list of expressions

pmer committed:
Add bare expr

pmer committed:
grammar: Add unit type

pmer committed:
Rename NullVal to UnitVal

pmer committed:
ExprBuilder: Add let visitLet

pmer committed:
ExprBuilder: Add value-literal expressions

pmer committed:
ExprBuilder: Add function-decl expressions

pmer committed:
ExprBuilder: Add binary operator expressions

pmer committed:
ExprBuilder: Add bare (sequence) expressions

pmer committed:
grammar: Introduce seq expression

- Meant to overcome lack of statements in this expression-oriented
  language
- Remove support for blank statements

pmer committed:
Add seq expression

pmer committed:
Remove while expression

pmer committed:
Add if expression

pmer committed:
grammar: Let expressions need types

pmer committed:
scripts: Write first Stratagem script

pmer committed:
Add classes for Types

pmer committed:
ExprBuilder: Add function-decl expressions

pmer committed:
ExprBuilder: Add function-app expressions

pmer committed:
ExprBuilder: Remove parens expressions

pmer committed:
ExprBuilder: Remove block expressions

Since we now use seq expressions

pmer committed:
ExprBuilder: Update syntax for Java 8

pmer committed:
Makefile: Run Stratagem scripts

pmer committed:
ExpressionTest: Remove extra whitespace

pmer committed:
Expression: Add SeqExpr(Expression[]) constructor

pmer committed:
ExpressionTest: Update tests for new SeqExpr

pmer committed:
ExpressionTest: Update syntax for Java 8

pmer committed:
Expression: Remove PrintExpr

pmer committed:
grammar: Typo

pmer committed:
Expression: Remove AssignExpr

pmer committed:
Expression: Remove WhileExpr

pmer committed:
Expression: Remove VarDeclExpr

pmer commented:
There's currently an error with the parsing because the Stratagem program:

```
let x: Int = 0 in x + 1;
let x: Int = 3 + 4 in x;
let y: Bool = true in ();
fn(z: Bool): Bool { z; z }(false)
```

yields the following output:

```
line 1:5 mismatched input ':' expecting '='
line 2:5 mismatched input ':' expecting '='
line 3:5 mismatched input ':' expecting '='
```

Any ideas?

pmer committed:
ExpressionTest: Remove AssignExpr test

pmer committed:
ExpressionTest: Update SeqExpr test

pmer committed:
ExpressionTest: Remove WhileExpr test

pmer committed:
Add Object[] constructors

For:

- FunctionDeclExpr
- FunctionAppExpr
- ClosureVal

pmer committed:
ExpressionTest: Update scope tests

We drop the test that exercises assignment statements.

pmer committed:
ExpressionTest: Remove VarDecl test

pmer committed:
grammar: Combine TYPE_UNIT and LIT_UNIT

Unit was always being lexed as TYPE_UNIT which meant that no LIT_UNITs
ever appears in the token stream.

pmer committed:
Interpreter: Print final value of program

pmer commented:
The issue was that `make run` didn't compile the sources and neither did my IDE when I modified a file and saved.

Thus, `make run` was always using an out-of-date build of the program.

taustin commented on Makefile:
```diff
@@ -1,15 +1,14 @@
+ANTLR_JAR=lib/antlr-4.6-complete.jar
 
 # These variables should not need to be changed
 GRAMMAR_NAME=FeatherweightJavaScript
```
Probably should change the name of the grammar file to Stratagem.g4

taustin commented on src/edu/sjsu/fwjs/Expression.java:
```diff
@@ -45,20 +43,6 @@ public Value evaluate(Environment env) {
-        this.exp = exp;
-    }
-    public Value evaluate(Environment env) {
-        Value v = exp.evaluate(env);
```
We'll need some way to write to stdout, although we can probably ignore it in our semantics.  Or maybe treat it as an operator?

pmer replied:
Hmm, it may be unavoidable to do this. Originally I was thinking it might be nice to make print be a function. This way it could be passed around as an argument to other higher-order functions. But since we don't have polymorphism we would need one print function for every type, which is untenable.

Therefore, let's add this back in.

pmer replied:
Done.

taustin commented on src/edu/sjsu/fwjs/Expression.java:
```diff
@@ -145,91 +131,43 @@ public Value evaluate(Environment env) {
-        return e2.evaluate(env);
-    }
-}
+    private static final Expression[] expressionArrayHint = new Expression[0];
```
I like this change -- I think it will be more efficient that the old approach.

taustin commented on src/edu/sjsu/fwjs/Type.java:
```diff
@@ -0,0 +1,91 @@
+package edu.sjsu.fwjs;
```
The class seems good, though we should probably change the package name to edu.sjsu.stratagem.

taustin commented on src/edu/sjsu/fwjs/Value.java:
```diff
@@ -83,23 +83,30 @@ public String toString() {
+        this.body = body;
+        this.outerEnv = env;
+    }
     public String toString() {
```
Hmm...  Perhaps we should add in typing annotations to the String representation as well.

pmer replied:
Good call. If you don't mind, I'll do this with the addition of the type checker since at that time we'll start remembering the type of a function (and therefore its closures). For now I've added a TODO.

taustin commented:
Hi Paul,

Very nice work.  I've approved and added a few minor comments.  See you 
on Wednesday.

--Austin

Paul Merrill wrote:
>
> The issue was that |make run| didn't compile the sources and neither 
> did my IDE when I modified a file and saved.
>
> Thus, |make run| was always using an out-of-date build of the program.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub 
> <https://github.com/taustin/stratagem/pull/2#issuecomment-285202252>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AAYpodC2uiFZuR5sY23oMwieuh3vTofjks5rjzdZgaJpZM4MXObm>.
>

pmer committed:
grammar: Rename grammar to Stratagem

pmer committed:
Rename source package to edu.sjsu.stratagem

pmer committed:
Value: Add comment about printing type annotations

pmer committed:
Add print expression

- Makes the Unit value a singleton
- Fixes strings lexing
- Alters Makefile to discover all scripts automatically

pmer committed:
Makefile: Add 'all' as dependency to 'run' target

Prevents running tests with out-of-date classes.

pmer committed:
Makefile: Avoid redundant compilation

Through the use of the timestamp files technique.

We now only generate and/or compile if the corresponding sources files
have been modified.

pmer commented:
I have modified the Makefile to compile the sources when executing the 'run' target.

To keep the 'run' target fast, I've also modified the Makefile to skip compilation of the sources when the sources haven't changed.
